zkvm/docs/zkvm-blockchain.md

```diff
@@ -0,0 +1,406 @@
+     all-zero string of 32-bytes
+   - `timestamp_ms`:
+     `timestamp_ms`
+   - `txroot`:
+     `txroot`
+   - `utxoroot`:
+     `utxoroot`
+   - `nonceroot`:
+     `nonceroot`
+   - `refscount`:
+     `refscount`
+5. Return a block with `header` set to `header` and an empty `txs` list.
+
+## Join existing network
+
+A new node starts here when joining a running network.
+It must either:
+- obtain all historical blocks, [applying](#apply-block) them one by one to reproduce the latest [blockchain state](#blockchain-state); or
+- obtain a recent copy of the blockchain state `state` and begin applying blocks beginning at `state.tipheader.height+1`.
+
+An obtained (as opposed to computed) blockchain state `state` may be partially validated by computing the [Merkle patricia root hash](#merkle-patricia-tree) of `state.utxos` using label `"utxos"` and verifying that it equals `state.header.utxoroot`.
+
+
+## Validate block
+
+Validating a block checks it for correctness outside the context of a particular [blockchain state](#blockchain-state).
+
+Additional correctness checks against a particular blockchain state happen during the [apply block](#apply-block) procedure.
+
+Inputs:
+- `block`,
+  the block to validate.
+- `prevheader`,
+  the header of the previous block.
+
+Output:
+- list of
+  [transaction logs](zkvm-spec.md#transaction-log),
+  one for each transaction in block.txs.
+
+Procedure:
+1. Verify `block.header.version >= prevheader.version`.
+2. Verify `block.header.height == prevheader.height+1`.
+3. Verify `block.header.previd` equals the
+   [block ID](#block-id)
+   of `prevheader`.
+4. Verify `block.header.timestamp_ms > prevheader.timestamp_ms`.
+5. Verify `block.header.refscount >= 0` and `block.header.refscount <= prevheader.refscount + 1`.
+6. Compute `txroot`, the [Merkle root hash](zkvm-spec.md#merkle-binary-tree) of `block.txs`, using label `"transactions"`.
+7. Verify `txroot == block.header.txroot`.
+8. For each transaction `tx` in block.txs:
+   1. Verify `tx.mintimems == 0` or `tx.mintimems >= block.header.timestamp_ms`.
+   2. Verify `tx.maxtimems == 0` or `tx.maxtimems < block.header.timestamp_ms`.
+   3. If `block.header.version == 1`,
+      verify `tx.version == 1`.
+   4. [Execute](zkvm-spec.md#vm-execution)
+      `tx` to produce transaction log `txlog`.
+   5. Add `txlog` to the list of output logs.
+
+## Apply block
+
+Applying a block causes a node to replace its [blockchain state](#blockchain-state) with the updated state that results.
+
+Inputs:
+- `block`,
+  the block to apply.
+- `state`,
+  the current
+  blockchain state.
+
+Output:
+- New blockchain state.
+
+Procedure:
+1. [Validate](#validate-block)
+   `block` with `prevheader` set to `state.tipheader`.
+2. Let `state′` be `state`.
+3. Remove items from `state′.nonces` where the expiration timestamp is earlier than `block.header.timestamp_ms`.
+4. For each txlog from the validation step,
+   in order:
+   1. [Apply](#apply-transaction-log)
```

Comment:
maybe make the whole phrase "Apply txlog" a xref so it matches the title of that section better?